File: apps/worker-node/status.ts
import { WebsiteStatus, prisma } from "@repo/db/client";
import { getPreviousStatusRedis, setCurrentStatus  } from "@repo/redis";

// Define status decision logic
export const getWebsiteStatus = (statusCode : number | null) : WebsiteStatus => {
    if(statusCode === null) return WebsiteStatus.unknown
    else if(statusCode >= 200 && statusCode < 400) return WebsiteStatus.up
    return WebsiteStatus.down
}

export const getPreviousStatus = async (websiteId : string)  => {
    // 1. First Try Redis
    const cachedStatus = await getPreviousStatusRedis(websiteId);
    if(cachedStatus !== null){
        return cachedStatus;
    }
    console.log(cachedStatus);
    // 2. Fallback to DB ( Get details for website with id : websiteId )
    const website = await prisma.website.findUnique({
        where : {
            id : websiteId
        },
        select : {
            currentStatus : true
        }
    })
    console.log(website?.currentStatus);
    if(website?.currentStatus){
        
        await setCurrentStatus(websiteId,website.currentStatus);
        return website.currentStatus;
    }
    return null;
}

File: apps/api-go/internal/workers/monitoring-worker/status.go
package monitoringworker

import (
	"context"
	"database/sql"
	"log"

	"github.com/RitikaxG/runState/apps/api-go/internal/domain"
)

func (mw *MonitoringWorker) GetWebsiteStatus(statusCode int) domain.WebsiteStatus {

	if statusCode >= 200 && statusCode < 400 {
		return domain.WebsiteUp
	}
	return domain.WebsiteDown
}

func (mw *MonitoringWorker) GetPreviousStatus(
	ctx context.Context,
	websiteId string,
) (*domain.WebsiteStatus, error) {

	// // 1. Try Redis First
	// cachedStatus, err := mw.redis.GetPreviousStatusRedis(ctx, websiteId)
	// if err != nil {
	// 	return nil, err
	// }

	// // 2. If redis has the status , return immediately
	// if cachedStatus != nil {
	// 	return cachedStatus, nil
	// }

	// 3. Redis miss -> Fallback to db
	website, err := mw.websiteRepo.GetByID(ctx, websiteId)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil // website gone; ignore
		}
		return nil, err
	}

	if website.CurrentStatus == nil {
		log.Println("Previous Status: <nil>")
	} else {
		log.Println("Previous Status:", *website.CurrentStatus)
	}

	// If DB has status -> warm redis cache
	if website.CurrentStatus != nil {
		_ = mw.redis.SetCurrentStatus(ctx, websiteId, *website.CurrentStatus)
		return website.CurrentStatus, nil
	}

	return nil, nil
}
